tests/src/python/provider_python.py

```diff
@@ -0,0 +1,417 @@
+# -*- coding: utf-8 -*-
+"""QGIS python layer provider test.
+
+This module is a Python implementation of (a clone of) the core memory
+vector layer provider, to be used for test_provider_python.py
+
+.. note:: This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation; either version 2 of the License, or
+(at your option) any later version.
+"""
+__author__ = 'Alessandro Pasotti'
+__date__ = '2018-03-18'
+__copyright__ = 'Copyright 2018, The QGIS Project'
+# This will get replaced with a git SHA1 when you do a git archive
+__revision__ = '$Format:%H$'
+
+
+from qgis.core import (
+    QgsField,
+    QgsFields,
+    QgsLayerDefinition,
+    QgsPointXY,
+    QgsReadWriteContext,
+    QgsVectorLayer,
+    QgsFeatureRequest,
+    QgsFeature,
+    QgsGeometry,
+    QgsProject,
+    QgsWkbTypes,
+    QgsExpression,
+    QgsExpressionContext,
+    QgsExpressionContextUtils,
+    NULL,
+    QgsCoordinateTransform,
+    QgsMemoryProviderUtils,
+    QgsCoordinateReferenceSystem,
+    QgsRectangle,
+    QgsTestUtils,
+    QgsVectorDataProvider,
+    QgsAbstractFeatureSource,
+    QgsAbstractFeatureIterator,
+    QgsFeatureIterator,
+    QgsApplication,
+    QgsProviderRegistry,
+    QgsProviderMetadata,
+    QgsGeometryEngine,
+    QgsSpatialIndex,
+    QgsDataProvider,
+)
+
+from qgis.PyQt.QtCore import QVariant
+
+
+class PyFeatureIterator(QgsAbstractFeatureIterator):
+
+    def __init__(self, source, request):
+        super().__init__(request)
+        self._request = request if request is not None else QgsFeatureRequest()
+        self._source = source
+        self._index = 0
+        self._transform = QgsCoordinateTransform()
+        if self._request.destinationCrs().isValid() and self._request.destinationCrs() != self._source._provider.crs():
+            self._transform = QgsCoordinateTransform(self._source._provider.crs(), self._request.destinationCrs(), self._request.transformContext())
+        self._filter_rect = self.filterRectToSourceCrs(self._transform)
```

Comment:
Here's why the test is failing on Travis- you need to wrap this in a try block:


    try:
        self._filter_rect = self.filterRectToSourceCrs(self._transform)
    except QgsCsException as e:
        self.close()
        return

Reply:
Yes - it's very tricky to get the cs exceptions thrown consistently. They seem to be very platform dependent.